src/main.js

```diff
@@ -110,45 +115,317 @@ const renderFilmCard = (count, area) => {
     };
-    let filmCard = new CreateFilmCard(filmExample);
-    let filmPopupElement = new FilmPopup(filmExample);
+    mockArray.push(filmExample);
+  }
+  return mockArray;
+};
+
+const mockData = mockDate(8);
+
+let filmCards = [];
+let filmPopupCards = [];
+
+const createCardsData = (data) => {
+  filmCards = [];
+  filmPopupCards = [];
+  for (let i = 0; i < data.length; i++) {
+    filmCards[i] = new Film(data[i]);
+    filmPopupCards[i] = new FilmPopup(data[i]);
+  }
+};
+
+createCardsData(mockData);
+
+
+const updateFiltersData = (filterId, state) => {
+  const filterIndex = filtersData.findIndex((it) => it.id === filterId);
+  if (state) {
+    filtersData[filterIndex].amount++;
+  } else {
+    filtersData[filterIndex].amount--;
+  }
+};
+
+const renderFilmCard = (data, area) => {
+  for (let i = 0; i < filmCards.length; i++) {
+    let filmCard = filmCards[i];
+    let filmPopupElement = filmPopupCards[i];
+    let dataId = null;
+
     area.appendChild(filmCard.render());
 
-    filmCard.onComments = () => {
+    filmCard.onComments = (id) => {
+      dataId = id;
       document.body.appendChild(filmPopupElement.render());
     };
 
+    filmCard.onMarkAsWatched = (state, id) => {
+      dataId = id;
+      data[dataId].isAlreadyWatched = state;
+
+      updateFiltersData(`HistoryFilms`, state);
+
+      filmPopupElement.update(data[dataId]);
+      filmCard.reRender();
+
+      removeFilters(filters);
+      renderFilters(filtersData);
+      getStatsData();
+    };
+
+    filmCard.onAddToFavorite = (state, id) => {
+      dataId = id;
+      data[dataId].isFavorite = state;
+
+      updateFiltersData(`FavoritesFilms`, state);
+
+      filmPopupElement.update(data[dataId]);
+      filmCard.reRender();
+
+      removeFilters(filters);
+      renderFilters(filtersData);
+    };
+
+    filmCard.onAddToWatchList = (state, id) => {
+      dataId = id;
+      data[dataId].isWatchList = state;
+
+      updateFiltersData(`WatchlistFilms`, state);
+
+      filmPopupElement.update(data[dataId]);
+      filmCard.reRender();
+
+      removeFilters(filters);
+      renderFilters(filtersData);
+    };
+
     filmPopupElement.onClose = () => {
       filmPopupElement.unrender();
     };
-    filmPopupElement.onSubmit = (newObject) => {
-      filmExample.isAlreadyWatched = newObject.isAlreadyWatched;
-      filmExample.isFavorite = newObject.isFavorite;
-      filmExample.isWatchList = newObject.isWatchList;
-      filmCard.update(filmExample);
+
+    filmPopupElement.onSubmit = (newObject, id) => {
+      dataId = id;
+      const watchedState = data[dataId].isAlreadyWatched;
+      const favoriteState = data[dataId].isFavorite;
+      const watchlistState = data[dataId].isWatchList;
+
+      data[dataId].isAlreadyWatched = newObject.isAlreadyWatched;
+      data[dataId].isFavorite = newObject.isFavorite;
+      data[dataId].isWatchList = newObject.isWatchList;
+
+      if (!watchedState === data[dataId].isAlreadyWatched) {
```

Comment:
Кажется `watchedState`, `favoriteState`, и, `watchlistState`, больше запутывают, так же будет очевиднее:
```suggestion
      if (newObject.isAlreadyWatched !== data[dataId].isAlreadyWatched) {
```